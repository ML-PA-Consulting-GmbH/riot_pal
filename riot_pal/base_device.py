# Copyright (c) 2018 Kevin Weiss, for HAW Hamburg  <kevin.weiss@haw-hamburg.de>
#
# This file is subject to the terms and conditions of the MIT License. See the
# file LICENSE in the top level directory for more details.
# SPDX-License-Identifier:    MIT
"""Abstract Base Device for RIOT PAL
This module is the abstract device that interfaces to a driver.  It handles
the selection and initialzation of the driver.

Example:
    class ApplicationDevice(BaseDevice):
        def show_example(self):
            self._write('Send Example')
            print(self._read())

    example_use_case = ApplicationDevice('serial', port='my/port/name')
    example_use_case.show_example()
"""
import logging
try:
    from . import driver_manager
except SystemError:
    import driver_manager


class BaseDevice:
    """Instance for devices to connect and utilize drivers.

    Args:
        *args: Variable length argument list.
        **kwargs: Arbitrary keyword arguments.
    """

    def __init__(self, *args, **kwargs):
        self._driver = driver_manager.driver_from_config(*args, **kwargs)

    def close(self):
        """Closes the device connection."""
        self._driver.close()

    def _read(self):
        """Reads data from the driver.

        Returns:
            str: string of data if success, ERR string if failed
        """
        return self._driver.read()

    def _write(self, data):
        """Writes data to the driver.

        Args:
            data(str): Variable length argument list.
        """
        return self._driver.write(data)

    def is_connected_to_board(self):
        """Dummy - confirm if a connection is for target board."""
        logging.warning("Check if board is connected: dummy should be"
                        " implmeneted in subclasses")
        raise NotImplementedError()

    @classmethod
    def from_autodetect(cls, *args, **dev_config):
        """Connects to a range of possible configurations."""
        configs = driver_manager.available_configs(*args, **dev_config)
        logging.debug("Configs: %r", configs)
        for config in configs:
            for retry in range(0, 2):
                logging.debug("Autodetect attempt: %d", retry)
                conn = cls(**config)
                try:
                    if conn.is_connected_to_board():
                        return conn
                except Exception as err:
                    logging.debug("Cannot connect: %r", err)
                conn.close()

        raise ValueError("Could not locate board, check if board is"
                         "connected or is_connected_to_board is correct")

    @classmethod
    def copy_driver(cls, device):
        """Copies the driver instance so many devices can use one driver."""
        logging.debug("Cloning Driver: %r", device._driver)
        return cls(dev_type='driver', driver=device._driver)


def main():
    """Tests basic usage of the class

    Used for unit testing, information should be confirm with DEBUG info.
    """
    logging.getLogger().setLevel(logging.DEBUG)
    BaseDevice()


if __name__ == "__main__":
    main()
