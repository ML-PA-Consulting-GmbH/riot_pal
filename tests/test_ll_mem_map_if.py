"""Tests all functions supported functions."""
from pprint import pformat
from pathlib import Path
from serial import SerialException
from riot_pal import LLMemMapIf
import helpers


def try_connect(mem_map='mem_map.csv'):
    try:
        return LLMemMapIf(mem_map, 'serial', '/dev/ttyACM0', timeout=1)
    except SerialException:
        return LLMemMapIf(mem_map, 'serial', '/dev/ttyUSB0', timeout=1)


def test_philip_mem_map_basic(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.cmd_list) + '\n')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')
    regtest.write(pformat(mmif.execute_changes()) + '\n')
    mmif.close()


def test_philip_mem_map_read_reg(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')

    for cmd in sorted(mmif.cmd_list):
        # Supress unused variable warning
        cmd_vals = mmif.cmd_list[cmd]
        read_val = mmif.read_reg(cmd)
        if 'DEVICE_SPECIFIC' in cmd_vals['flag']:
            read_val.pop('data', None)
        elif cmd_vals['default'] is not '':
            assert helpers.weak_cmp(read_val['data'], cmd_vals['default'])
        regtest.write(pformat(read_val) + '\n')
    mmif.close()


def test_philip_mem_map_write_reg(regtest):
    mmif = try_connect(str(Path(__file__).parents[1]) + '/mem_map/philip_mem_map.csv')
    regtest.write(pformat(mmif.reset_mcu()) + '\n')

    for cmd in sorted(mmif.cmd_list):
        # Supress unused variable warning
        read_val = mmif.read_reg(cmd)
        write_ret = mmif.write_reg(cmd, helpers.try_add(read_val['data']))
        read_val = mmif.read_reg(cmd)
        regtest.write(pformat(write_ret) + '\n')
        regtest.write(pformat(read_val) + '\n')
    mmif.close()
