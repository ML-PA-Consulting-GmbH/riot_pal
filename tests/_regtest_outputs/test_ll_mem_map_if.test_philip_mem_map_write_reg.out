{'cmd': 'mcu_rst',
 'data': [' Initialized', ' Build Date: Aug 23 2018 09:57:12'],
 'msg': 'EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 89 1 0',
 'data': None,
 'msg': 'cmd=adc.0.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 89 2',
 'data': 1,
 'msg': 'cmd=adc.0.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 88 1',
 'data': None,
 'msg': 'cmd=adc.0.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 88 1',
 'data': 1,
 'msg': 'cmd=adc.0.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 96 0 0 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=adc.0.res.8 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 96 8',
 'data': ['0x0000000000000000'],
 'msg': 'cmd=adc.0.res.8 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 91 1',
 'data': None,
 'msg': 'cmd=adc.0.sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 91 1',
 'data': 1,
 'msg': 'cmd=adc.0.sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 92 1 0 0 0',
 'data': None,
 'msg': 'cmd=adc.0.value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 92 4',
 'data': 1,
 'msg': 'cmd=adc.0.value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 105 1 0',
 'data': None,
 'msg': 'cmd=adc.1.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 105 2',
 'data': 1,
 'msg': 'cmd=adc.1.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 104 1',
 'data': None,
 'msg': 'cmd=adc.1.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 104 1',
 'data': 1,
 'msg': 'cmd=adc.1.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 112 0 0 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=adc.1.res.8 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 112 8',
 'data': ['0x0000000000000000'],
 'msg': 'cmd=adc.1.res.8 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 107 1',
 'data': None,
 'msg': 'cmd=adc.1.sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 107 1',
 'data': 1,
 'msg': 'cmd=adc.1.sample_rate response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 108 1 0 0 0',
 'data': None,
 'msg': 'cmd=adc.1.value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 108 4',
 'data': 1,
 'msg': 'cmd=adc.1.value response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 34 1 0',
 'data': None,
 'msg': 'cmd=i2c.clk_stretch_delay response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 34 2',
 'data': 1,
 'msg': 'cmd=i2c.clk_stretch_delay response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 1',
 'data': None,
 'msg': 'cmd=i2c.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 1,
 'msg': 'cmd=i2c.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 2',
 'data': None,
 'msg': 'cmd=i2c.mode.addr_10_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 2,
 'msg': 'cmd=i2c.mode.addr_10_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 3',
 'data': None,
 'msg': 'cmd=i2c.mode.general_call response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 3,
 'msg': 'cmd=i2c.mode.general_call response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 4',
 'data': None,
 'msg': 'cmd=i2c.mode.nack_data response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 4,
 'msg': 'cmd=i2c.mode.nack_data response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 5',
 'data': None,
 'msg': 'cmd=i2c.mode.no_clk_stretch response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 5,
 'msg': 'cmd=i2c.mode.no_clk_stretch response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 32 6',
 'data': None,
 'msg': 'cmd=i2c.mode.reg_16_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 32 1',
 'data': 6,
 'msg': 'cmd=i2c.mode.reg_16_bit response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 40 1',
 'data': None,
 'msg': 'cmd=i2c.r_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 40 1',
 'data': 1,
 'msg': 'cmd=i2c.r_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 42 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=i2c.res.6 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 42 6',
 'data': ['0x000000000000'],
 'msg': 'cmd=i2c.res.6 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 36 86 0',
 'data': None,
 'msg': 'cmd=i2c.slave_addr_1 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 36 2',
 'data': 86,
 'msg': 'cmd=i2c.slave_addr_1 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 38 67 0',
 'data': None,
 'msg': 'cmd=i2c.slave_addr_2 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 38 2',
 'data': 67,
 'msg': 'cmd=i2c.slave_addr_2 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 1',
 'data': None,
 'msg': 'cmd=i2c.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 1,
 'msg': 'cmd=i2c.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 2',
 'data': None,
 'msg': 'cmd=i2c.status.AF response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 2,
 'msg': 'cmd=i2c.status.AF response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 3',
 'data': None,
 'msg': 'cmd=i2c.status.BERR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 3,
 'msg': 'cmd=i2c.status.BERR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 4',
 'data': None,
 'msg': 'cmd=i2c.status.BUSY response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 4,
 'msg': 'cmd=i2c.status.BUSY response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 5',
 'data': None,
 'msg': 'cmd=i2c.status.GENCALL response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 5,
 'msg': 'cmd=i2c.status.GENCALL response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 6',
 'data': None,
 'msg': 'cmd=i2c.status.OVR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 6,
 'msg': 'cmd=i2c.status.OVR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 33 7',
 'data': None,
 'msg': 'cmd=i2c.status.RSR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 33 1',
 'data': 7,
 'msg': 'cmd=i2c.status.RSR response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 41 1',
 'data': None,
 'msg': 'cmd=i2c.w_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 41 1',
 'data': 1,
 'msg': 'cmd=i2c.w_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 123 1',
 'data': None,
 'msg': 'cmd=pwm.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 123 1',
 'data': 1,
 'msg': 'cmd=pwm.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 121 1 0',
 'data': None,
 'msg': 'cmd=pwm.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 121 2',
 'data': 1,
 'msg': 'cmd=pwm.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 124 1 0 0 0',
 'data': None,
 'msg': 'cmd=pwm.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 124 4',
 'data': 1,
 'msg': 'cmd=pwm.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 120 1',
 'data': None,
 'msg': 'cmd=pwm.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 120 1',
 'data': 1,
 'msg': 'cmd=pwm.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 128 0 0 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=pwm.res.8 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 128 8',
 'data': ['0x0000000000000000'],
 'msg': 'cmd=pwm.res.8 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 216 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 '
        '0 0 0 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=res.40 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 216 40',
 'data': ['0x00000000000000000000000000000000000000000000000000000000000000000000000000000000'],
 'msg': 'cmd=res.40 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 83 1',
 'data': None,
 'msg': 'cmd=rtc.day_of_month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 83 1',
 'data': 1,
 'msg': 'cmd=rtc.day_of_month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 84 1',
 'data': None,
 'msg': 'cmd=rtc.day_of_week response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 84 1',
 'data': 1,
 'msg': 'cmd=rtc.day_of_week response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 82 1',
 'data': None,
 'msg': 'cmd=rtc.hour response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 82 1',
 'data': 1,
 'msg': 'cmd=rtc.hour response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 81 1',
 'data': None,
 'msg': 'cmd=rtc.minute response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 81 1',
 'data': 1,
 'msg': 'cmd=rtc.minute response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 85 1',
 'data': None,
 'msg': 'cmd=rtc.month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 85 1',
 'data': 1,
 'msg': 'cmd=rtc.month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 87 1',
 'data': 13,
 'msg': 'cmd=rtc.res.1 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 87 1',
 'data': 0,
 'msg': 'cmd=rtc.res.1 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 80 1',
 'data': None,
 'msg': 'cmd=rtc.second response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 80 1',
 'data': 1,
 'msg': 'cmd=rtc.second response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 86 1',
 'data': None,
 'msg': 'cmd=rtc.year response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 86 1',
 'data': 1,
 'msg': 'cmd=rtc.year response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 49 1 0 0 0',
 'data': None,
 'msg': 'cmd=spi.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 49 4',
 'data': 1,
 'msg': 'cmd=spi.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 48 1',
 'data': None,
 'msg': 'cmd=spi.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 48 1',
 'data': 1,
 'msg': 'cmd=spi.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 53 0 0 0 0 0 0 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=spi.res.11 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 53 11',
 'data': ['0x0000000000000000000000'],
 'msg': 'cmd=spi.res.11 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 19 24',
 'data': 13,
 'msg': 'cmd=sys.build_time.day_of_month response=EACCES-Permission denied '
        '[13]',
 'result': 'Error'}
{'cmd': 'rr 19 1',
 'data': 23,
 'msg': 'cmd=sys.build_time.day_of_month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 20 1',
 'data': 13,
 'msg': 'cmd=sys.build_time.day_of_week response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 20 1',
 'data': 0,
 'msg': 'cmd=sys.build_time.day_of_week response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 18 10',
 'data': 13,
 'msg': 'cmd=sys.build_time.hour response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 18 1',
 'data': 9,
 'msg': 'cmd=sys.build_time.hour response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 17 58',
 'data': 13,
 'msg': 'cmd=sys.build_time.minute response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 17 1',
 'data': 57,
 'msg': 'cmd=sys.build_time.minute response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 21 9',
 'data': 13,
 'msg': 'cmd=sys.build_time.month response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 21 1',
 'data': 8,
 'msg': 'cmd=sys.build_time.month response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 23 1',
 'data': 13,
 'msg': 'cmd=sys.build_time.res.1 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 23 1',
 'data': 0,
 'msg': 'cmd=sys.build_time.res.1 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 16 13',
 'data': 13,
 'msg': 'cmd=sys.build_time.second response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 16 1',
 'data': 12,
 'msg': 'cmd=sys.build_time.second response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 22 19',
 'data': 13,
 'msg': 'cmd=sys.build_time.year response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 22 1',
 'data': 18,
 'msg': 'cmd=sys.build_time.year response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 28 1',
 'data': None,
 'msg': 'cmd=sys.cr response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 28 1',
 'data': 1,
 'msg': 'cmd=sys.cr response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 28 2',
 'data': None,
 'msg': 'cmd=sys.cr.dut_rst response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 28 1',
 'data': 2,
 'msg': 'cmd=sys.cr.dut_rst response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 24 166 66 0 0',
 'data': None,
 'msg': 'cmd=sys.device_num response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 24 4',
 'data': 17062,
 'msg': 'cmd=sys.device_num response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 12 26 39 0 0',
 'data': 13,
 'msg': 'cmd=sys.fw_rev response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 12 4',
 'data': 10009,
 'msg': 'cmd=sys.fw_rev response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 29 1 0 0 0',
 'data': 13,
 'msg': 'cmd=sys.res.3 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 29 3',
 'data': 0,
 'msg': 'cmd=sys.res.3 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 0 50 255 110 6 77 77 50 56 16 67 22 67',
 'data': 13,
 'msg': 'cmd=sys.sn.12 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 0 12',
 'data': ['0x4316431038324D4D066EFF32'],
 'msg': 'cmd=sys.sn.12 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 139 1',
 'data': None,
 'msg': 'cmd=tmr.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 139 1',
 'data': 1,
 'msg': 'cmd=tmr.duty response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 137 1 0',
 'data': None,
 'msg': 'cmd=tmr.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 137 2',
 'data': 1,
 'msg': 'cmd=tmr.error_code response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 140 1 0 0 0',
 'data': None,
 'msg': 'cmd=tmr.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 140 4',
 'data': 1,
 'msg': 'cmd=tmr.freq response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 144 1 0 0 0',
 'data': None,
 'msg': 'cmd=tmr.hi_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 144 4',
 'data': 1,
 'msg': 'cmd=tmr.hi_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 148 1 0 0 0',
 'data': None,
 'msg': 'cmd=tmr.lo_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 148 4',
 'data': 1,
 'msg': 'cmd=tmr.lo_us response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 136 1',
 'data': None,
 'msg': 'cmd=tmr.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 136 1',
 'data': 1,
 'msg': 'cmd=tmr.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 65 1 194 1 0',
 'data': None,
 'msg': 'cmd=uart.baud response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 65 4',
 'data': 115201,
 'msg': 'cmd=uart.baud response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 73 1',
 'data': None,
 'msg': 'cmd=uart.ctrl response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 73 1',
 'data': 1,
 'msg': 'cmd=uart.ctrl response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 73 2',
 'data': None,
 'msg': 'cmd=uart.ctrl.parity response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 73 1',
 'data': 2,
 'msg': 'cmd=uart.ctrl.parity response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 73 3',
 'data': None,
 'msg': 'cmd=uart.ctrl.rts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 73 1',
 'data': 3,
 'msg': 'cmd=uart.ctrl.rts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 73 4',
 'data': None,
 'msg': 'cmd=uart.ctrl.stop_bits response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 73 1',
 'data': 4,
 'msg': 'cmd=uart.ctrl.stop_bits response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 64 1',
 'data': None,
 'msg': 'cmd=uart.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 64 1',
 'data': 1,
 'msg': 'cmd=uart.mode response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 75 0 0 0 0 0',
 'data': 13,
 'msg': 'cmd=uart.res.5 response=EACCES-Permission denied [13]',
 'result': 'Error'}
{'cmd': 'rr 75 5',
 'data': ['0x0000000000'],
 'msg': 'cmd=uart.res.5 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 69 1 0',
 'data': None,
 'msg': 'cmd=uart.rx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 69 2',
 'data': 1,
 'msg': 'cmd=uart.rx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 74 1',
 'data': None,
 'msg': 'cmd=uart.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 74 1',
 'data': 1,
 'msg': 'cmd=uart.status response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 74 2',
 'data': None,
 'msg': 'cmd=uart.status.cts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 74 1',
 'data': 2,
 'msg': 'cmd=uart.status.cts response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 71 1 0',
 'data': None,
 'msg': 'cmd=uart.tx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 71 2',
 'data': 1,
 'msg': 'cmd=uart.tx_count response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'wr 152 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 '
        '24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 '
        '47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63',
 'data': None,
 'msg': 'cmd=user_reg.64 response=EOK-command success [0]',
 'result': 'Success'}
{'cmd': 'rr 152 64',
 'data': ['0x3F3E3D3C3B3A393837363534333231302F2E2D2C2B2A292827262524232221201F1E1D1C1B1A191817161514131211100F0E0D0C0B0A09080706050403020100'],
 'msg': 'cmd=user_reg.64 response=EOK-command success [0]',
 'result': 'Success'}
